### PR TITLE
Reland "[Windows] Move to FlutterCompositor for rendering"

### PR DESCRIPTION
**Pull request by loic-sharma**
## Original pull request description

This migrates the Windows embedder to `FlutterCompositor` so that the engine renders off-screen to a framebuffer instead of directly onto the window's surface. This will allow us to support platform views and multiple views on Windows.

Addresses https://github.com/flutter/flutter/issues/128904

## Reland (again)

https://github.com/flutter/engine/pull/49262 was reverted as it regressed [`package:material_floating_search_bar_2`](https://pub.dev/packages/material_floating_search_bar_2/versions/0.5.0). See: https://github.com/flutter/flutter/issues/140828

This pull request is split into the following commits:

1. https://github.com/flutter/engine/pull/49726/commits/d337378a74c1c46ff10d0dcf9974c07f2408c524 is the previous reland pull request, unchanged
2. https://github.com/flutter/engine/pull/49726/commits/e866af02296cec6e451e2e455fe7e260325c4593 disables the scissor test before blitting the framebuffer, allowing us to "force" copy the framebuffer's contents by ignoring scissoring values

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/compositor_opengl.cc [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+                       GL_COLOR_BUFFER_BIT,  // mask
+                       GL_NEAREST            // filter
+  );
+
+  gl_->BindFramebuffer(GL_READ_FRAMEBUFFER, 0);
+  gl_->BindFramebuffer(GL_DRAW_FRAMEBUFFER, 0);
+
+  gl_->Enable(GL_SCISSOR_TEST);
+  gl_->Enable(GL_DEPTH_TEST);
+  gl_->Enable(GL_STENCIL_TEST);
```
Comment (loic-sharma) [inline shell/platform/windows/compositor_opengl.cc:161]: This is new:

1. Re-enable scissor, depth, and stencil tests
1. Previously we were unbinding the read/draw framebuffers. This doesn't appear to be strictly necessary, but from my understanding being explicit may prevent future bugs

**Review by jonahwilliams — COMMENTED** [review]
Not that it should block this change, but does Impeller still work with this configuration?

**Comment by loic-sharma** [comment]
@jonahwilliams It does not. Impeller invokes [`FlutterOpenGLRendererConfig.fbo_with_frame_info_callback`](https://github.com/flutter/engine/blob/42738dd0b259deee0d9245f3918e27ec7afc598e/shell/platform/embedder/embedder.h#L565), which now immediately exists on Windows since the engine should be rendering to a backing store instead of the window's surface

**Comment by jonahwilliams** [comment]
Thanks for checking! Can you file a bug for us with some information on what we need to update to make it work again?

File: impeller/renderer/backend/gles/blit_command_gles.cc [hunk 2]
```diff
@@ -48,7 +48,7 @@ static std::optional<GLuint> ConfigureFBO(
   }
 
   if (gl.CheckFramebufferStatus(fbo_type) != GL_FRAMEBUFFER_COMPLETE) {
-    VALIDATION_LOG << "Could not create a complete frambuffer.";
+    VALIDATION_LOG << "Could not create a complete framebuffer.";
```
Comment (cbracken) [inline impeller/renderer/backend/gles/blit_command_gles.cc:51]: frambuffer: most, but not quite all, of a framebuffer. 🤣

File: shell/platform/linux/fl_backing_store_provider.cc [hunk 3]
```diff
@@ -82,7 +82,8 @@ uint32_t fl_backing_store_provider_get_gl_format(FlBackingStoreProvider* self) {
   // In Linux kN32_SkColorType is assumed to be kBGRA_8888_SkColorType.
   // So we must choose a valid gl format to be compatible with surface format
   // BGRA8.
-  // Following logics are copied from Skia GrGLCaps.cpp.
+  // Following logics are copied from Skia GrGLCaps.cpp:
```
Comment (cbracken) [inline shell/platform/linux/fl_backing_store_provider.cc:85]: Entirely non-critical but may as well fix it while you're in here: "Following logic is copied"

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

**Review by yaakovschectman — APPROVED** [review]

File: shell/platform/linux/fl_backing_store_provider.cc (hunk 3 above)
Comment (loic-sharma) [reply shell/platform/linux/fl_backing_store_provider.cc]: Fixed!

**Comment by azchohfi** [comment]
Would this break Windows 7 compatibility, or is this being used only if Win8+?
We might need to wait for https://github.com/flutter/flutter/issues/140830

**Comment by loic-sharma** [comment]
@azchohfi For now we still fully support Windows 7. This PR doesn't use DirectComposition - is that your concern?

**Comment by loic-sharma** [comment]
@jonahwilliams Sure! Opened: https://github.com/flutter/flutter/issues/141482.

**Comment by azchohfi** [comment]
> @azchohfi For now we still fully support Windows 7. This PR doesn't use DirectComposition - is that your concern?

I completely misread the PR, ignore my comment. My bad.

File: shell/platform/windows/compositor_opengl.cc [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+  if (!engine_->surface_manager()->MakeCurrent()) {
+    return false;
+  }
+
+  auto source_id = layers[0]->backing_store->open_gl.framebuffer.name;
+
+  gl_->Disable(GL_SCISSOR_TEST);
+  gl_->Disable(GL_DEPTH_TEST);
+  gl_->Disable(GL_STENCIL_TEST);
```
[the remark this answers is not in the document]
Comment (bdero) [reply shell/platform/windows/compositor_opengl.cc:139]: I don't think you need to worry about the depth/stencil state here. The [4.6 spec](https://registry.khronos.org/OpenGL/specs/gl/glspec46.core.pdf) says:
```
When values are written to the draw buffers, blit operations bypass most of the
fragment pipeline. The only fragment operations which affect a blit are the pixel
ownership test, the scissor test, and sRGB conversion (see section 17.3.7). Color,
depth, and stencil masks (see section 17.4.2) are ignored.
```

File: shell/platform/windows/compositor_opengl.cc (hunk 1 above)
Comment (bdero) [reply shell/platform/windows/compositor_opengl.cc:161]: Usually GL renderers will set all necessary state before performing a set of relevant operations. Doubly so if you're disabling a feature. So I'd skip this attempt at restoring?

File: shell/platform/windows/compositor_opengl.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc:139]: Looks like you're right: the fix still works if we disable only the scissor test. Will update!

File: shell/platform/windows/compositor_opengl.cc (hunk 1 above)
Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc]: From an offline conversation with Brandon:

> GL_SCISSOR_TEST is off by default, and if it's needed for some other framebuffer operation in the future, code that intends to use the scissor test should be explicitly turning it on just before it's needed anyhow.

Comment (loic-sharma) [reply shell/platform/windows/compositor_opengl.cc]: Marking this as resolved as this restore logic was removed.

**Comment by knopp** [comment]
Just to be clear, this change introduced an extra blit that's always there? What exactly is the reason for it? Why not use the back framebuffer?

**Comment by loic-sharma** [comment]
@knopp and I talked on Discord, but I'll capture our conversation here as well in case other folks have the same question.

Currently the engine's backing stores are interchangeable if they have the same size. The intermediary textures plus extra blit allow us to keep this interchangeable model and also render to multiple views, though of course this results in higher memory usage.

We plan to have a conversation with the engine team to see if we can "tie" backing stores to a view. We expect this would require some minor embedder API changes. See: http://flutter.dev/go/backing-stores-for-multi-view-partial-repaint

**Pull request merged**